File: src/main/java/sparkstreaming_to_kafka_optimizer/broadcastKafkaProducer.java
package sparkstreaming_to_kafka_optimizer;

import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import java.util.Properties;
import java.util.concurrent.Future;
import java.io.Serializable;

public class broadcastKafkaProducer implements Serializable {

    private static KafkaProducer produecer;
    private Properties properties;

    private broadcastKafkaProducer(Properties properties){
        this.properties=properties;
    }

    public KafkaProducer getProduecer(){
        if(produecer == null){
            return new KafkaProducer<>(properties);
        }else{
            return produecer;
        }
    }

    public void send(String topic, String value) {
        produecer.send(new ProducerRecord(topic,value));
    }

    public Future send(String topic, String key, String value) {
        Future future = produecer.send(new ProducerRecord(topic, key, value));
        return future;
    }

}

File: src/main/scala/sparkstreaming_to_kafka/broadcastKafkaProducer.scala
package sparkstreaming_to_kafka

import java.util.concurrent.Future

import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord, RecordMetadata}

class broadcastKafkaProducer[K, V](createproducer: () => KafkaProducer[K, V]) extends Serializable {
  lazy val producer = createproducer()

  def send(topic: String, key: K, value: V): Future[RecordMetadata] = producer.send(new ProducerRecord[K, V](topic, key, value))

  def send(topic: String, value: V): Future[RecordMetadata] = producer.send(new ProducerRecord[K, V](topic, value))

}

object broadcastKafkaProducer {


  import scala.collection.JavaConversions._

  def apply[K, V](config: Map[String, Object]): broadcastKafkaProducer[K, V] = {
    val createProducerFunc = () => {
      val producer = new KafkaProducer[K, V](config)
      sys.addShutdownHook({producer.close()})
      producer
    }
    new broadcastKafkaProducer(createProducerFunc)
  }

  def apply[K, V](config: java.util.Properties): broadcastKafkaProducer[K, V] = apply(config.toMap)
}